### PR TITLE
Fix bug that made List and Dict unusable types on python 3.7 and 3.8

### LINKED ISSUE #325: Support for Optional[Dict] = None on Python 3.7, 3.8
The following piece of code works on python 3.6, but not python 3.7, nor python 3.8 on Ubuntu.
```python
from omegaconf import OmegaConf
from dataclasses import dataclass
from typing import Dict, Optional
import sys


@dataclass
class TestConfig:
    myDict: Optional[Dict] = None


print(sys.version)
schema = OmegaConf.structured(TestConfig)
```

python3.6 output
```
3.6.9 (default, Jul 17 2020, 12:50:27) 
[GCC 8.4.0]
```
python3.7 output
```
3.7.5 (default, Nov  7 2019, 10:50:52) 
[GCC 8.3.0]
Traceback (most recent call last):
  File "omconf_min.py", line 13, in <module>
    schema = OmegaConf.structured(TestConfig)
  File "/home/jasoli/py37_venv_test/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 136, in structured
    return OmegaConf.create(obj, parent)
  File "/home/jasoli/py37_venv_test/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 173, in create
    return OmegaConf._create_impl(obj=obj, parent=parent)
  File "/home/jasoli/py37_venv_test/lib/python3.7/site-packages/omegaconf/omegaconf.py", line 241, in _create_impl
    format_and_raise(node=None, key=None, value=None, msg=str(e), cause=e)
  File "/home/jasoli/py37_venv_test/lib/python3.7/site-packages/omegaconf/_utils.py", line 628, in format_and_raise
    _raise(ex, cause)
  File "/home/jasoli/py37_venv_test/lib/python3.7/site-packages/omegaconf/_utils.py", line 553, in _raise
    raise ex  # set end OC_CAUSE=1 for full backtrace
omegaconf.errors.ValidationError: Unsupported key type ~KT, Unsupported value type : ~VT
        full_key: 
        reference_type=None
        object_type=None
```
python3.8 output
```
3.8.0 (default, Oct 28 2019, 16:14:01) 
[GCC 8.3.0]
Traceback (most recent call last):
  File "omconf_min.py", line 13, in <module>
    schema = OmegaConf.structured(TestConfig)
  File "/home/jasoli/py38_venv_test/lib/python3.8/site-packages/omegaconf/omegaconf.py", line 133, in structured
    return OmegaConf.create(obj, parent)
  File "/home/jasoli/py38_venv_test/lib/python3.8/site-packages/omegaconf/omegaconf.py", line 170, in create
    return OmegaConf._create_impl(obj=obj, parent=parent)
  File "/home/jasoli/py38_venv_test/lib/python3.8/site-packages/omegaconf/omegaconf.py", line 238, in _create_impl
    format_and_raise(node=None, key=None, value=None, msg=str(e), cause=e)
  File "/home/jasoli/py38_venv_test/lib/python3.8/site-packages/omegaconf/_utils.py", line 590, in format_and_raise
    raise_(ex, cause)
  File "/home/jasoli/py38_venv_test/lib/python3.8/site-packages/omegaconf/_utils.py", line 515, in raise_
    raise ex
omegaconf.errors.ValidationError: Unsupported value type : ~VT
        full_key: 
        reference_type=None
        object_type=None
```

Can I ask why this is happening? And if a fix for python3.7 and 3.8 will be coming?

### PR DESCRIPTION
**Pull request by pereman2**
Closes #325 .

When getting arguments of a Dict or List varied in 3.6 and > 3.6. In 3.7 and 3.8 the __args__ for Dict and List are not None, instead, they are TypeVars.

I added `@pytest.fixture` in some test because switching to 3.7 raised an error when testing.

### CODE REVIEW

File: omegaconf/listconfig.py [hunk 1]
```diff
@@ -65,6 +65,12 @@ def __init__(
         except Exception as ex:
             format_and_raise(node=None, key=None, value=None, cause=ex, msg=str(ex))
 
+    def _resolve_generic_ref_type(self, ref_type: type) -> type:
+        if ref_type == List:
+            return List[Any]
+        else:
+            return ref_type
+
```
Comment (omry) [inline omegaconf/listconfig.py:73]: why is this not always List[element_type] ?

File: tests/test_utils.py [hunk 2]
```diff
@@ -183,6 +183,7 @@ def test_get_structured_config_data(test_cls_or_obj: Any, expectation: Any) -> N
         assert d["dict1"] == {}
 
 
+@pytest.fixture  # type: ignore
```
Comment (omry) [inline tests/test_utils.py:186]: what's with the @pytest.fixtures?

Comment (pereman2) [reply tests/test_utils.py:186]: I mentioned it above. When I switched to python 3.7 running pytest raised and error that said that a fixture was missing.

Comment (omry) [reply tests/test_utils.py:186]: How come it was not an issue before this diff?

Comment (pereman2) [reply tests/test_utils.py:186]: I've been on python 3.6 the whole time.

Comment (pereman2) [reply tests/test_utils.py:186]: Were you on a greater version ?

Comment (omry) [reply tests/test_utils.py:186]: 1. I am on 3.8
2. nox is testing 3.6, 3.7 and 3.8 in CI.

File: omegaconf/listconfig.py (hunk 1 above)
Comment (pereman2) [reply omegaconf/listconfig.py:73]: Yeah that would be better, the same goes to Dict. I'll change it.

Comment (omry) [reply omegaconf/listconfig.py:73]: I am not sure this is legal Python, but you get the idea.

File: tests/test_utils.py (hunk 2 above)
Comment (pereman2) [reply tests/test_utils.py:186]: Wait now it doesn't raise the error. My bad i don't know how did it happen

File: tests/test_basic_ops_dict.py [hunk 3]
```diff
@@ -111,7 +111,6 @@ def test_subscript_setitem_with_dot_warning() -> None:
         c = OmegaConf.create({"foo.bar": IntegerNode(10)})
 
 
-@pytest.fixture  # type: ignore
 def test_subscript_set_with_dot_warning_suppressed(recwarn: Any, mocker: Any) -> None:
     c = OmegaConf.create()
     mocker.patch.dict(os.environ, {"OC_DISABLE_DOT_ACCESS_WARNING": "1"})
```
Comment (omry) [inline tests/test_basic_ops_dict.py]: rebase, I removed this function.

File: omegaconf/omegaconf.py [hunk 4]
```diff
@@ -226,7 +226,10 @@ def _create_impl(  # noqa F811
                     ref_type = OmegaConf.get_type(obj)
```
Comment (omry) [inline omegaconf/omegaconf.py:226]: what is this actually doing?
what will ref_type be if is_primitive_list(obj)?
what will it be if OmegaConf.is_list(obj)?

File: tests/structured_conf/data/dataclasses.py [hunk 5]
```diff
@@ -475,3 +475,13 @@ class MissingStructuredConfigField:
 class ListClass:
     list: List[int] = field(default_factory=lambda: [])
     tuple: Tuple[int, int] = field(default_factory=lambda: (1, 2))
+
+
+@dataclass
+class GenericList:
+    list: Optional[List] = None  # type: ignore
```
Comment (omry) [inline tests/structured_conf/data/dataclasses.py:482]: Also test non optional list and dict.

File: omegaconf/omegaconf.py (hunk 4 above)
Comment (pereman2) [reply omegaconf/omegaconf.py:226]: My bad, I changed it because I was trying another thing.

Comment (omry) [reply omegaconf/omegaconf.py]: It's good to review your own code critically, making sure you can justify each change.

File: tests/structured_conf/data/dataclasses.py [hunk 6]
```diff
@@ -479,9 +479,19 @@ class ListClass:
...
 
 
 @dataclass
 class GenericDict:
+    dict: Dict = field(default_factory=lambda: {})  # type: ignore
+
+
+@dataclass
+class OptionalGenericList:
+    list: Optional[List] = None  # type: ignore
```
Comment (omry) [inline tests/structured_conf/data/dataclasses.py:492]: maybe:
```
GenericDict:
  dict: Dict = field(default_factory=lambda: {})
  opt_dict: Optional[Dict] = field(default_factory=lambda: {})
```

File: omegaconf/_utils.py [hunk 7]
```diff
@@ -465,7 +465,7 @@ def get_dict_key_value_types(ref_type: Any) -> Tuple[Any, Any]:
 
     key_type: Any
     element_type: Any
-    if ref_type is None:
+    if ref_type is None or ref_type == Dict:
```
Comment (omry) [inline omegaconf/_utils.py:468]: so we make the None here, and Union[str, Enum] and Any later.
seems like this can be simplified more returning (Union[str, Enum], Any) here in the first place.
(same for list).

Comment (pereman2) [reply omegaconf/_utils.py:468]: In order to simplify this as returning Union[str, Enum] and Any I have to change more logic in dictconfig and listconfig. I think for now it should be better this way.

Comment (omry) [reply omegaconf/_utils.py:468]: Did you push this?
I just spent a few minutes trying to figure that out and I don't think you did.

Comment (pereman2) [reply omegaconf/_utils.py:468]: No I didn't. I changed the test as you requested above and I was looking at changing this.

Comment (omry) [reply omegaconf/_utils.py:468]: Gotcha. send the update and we can discuss over the changes.

**Comment by omry** [comment]
re-request review when ready for review.

File: omegaconf/_utils.py (hunk 7 above)
Comment (pereman2) [reply omegaconf/_utils.py:468]: @omry I don't know if returning Union[str, Enum] and Any would simplify this now. It would mean changing more internal logic like changing ref_type to Any as default.

Comment (omry) [reply omegaconf/_utils.py:468]: besides the fact that it means changing code, isn't Any a better default than None?

Comment (pereman2) [reply omegaconf/_utils.py:468]: I think Any brings less problems than None as a default.

Comment (omry) [reply omegaconf/_utils.py:468]: Right, so are we in agreement that it's better to use Any as a default, not not using None and alternating back and forth everywhere.

We can make these changes in a followup.

Comment (pereman2) [reply omegaconf/_utils.py:468]: Perfect.

File: omegaconf/listconfig.py [hunk 8]
```diff
@@ -66,6 +66,14 @@ def __init__(
         except Exception as ex:
             format_and_raise(node=None, key=None, value=None, cause=ex, msg=str(ex))
 
+    def _resolve_generic_ref_type(
+        self, ref_type: Optional[Type[Any]], element_type: Optional[Type[Any]]
+    ) -> Optional[Type[Any]]:
+        if ref_type == List:
+            return List[element_type]  # type: ignore
```
Comment (omry) [inline omegaconf/listconfig.py:73]: Can you add a test that verifies that pickle works for Structured Configs?
I remember issues around serializing things like `List[int]`.

Comment (omry) [reply omegaconf/listconfig.py:73]: There should be some tests already but I am not sure they are covering this properly.

Comment (pereman2) [reply omegaconf/listconfig.py:73]: it looks like it doesn't work with picke, output: `Can't pickle typing.List[typing.Any]: it's not the same object as typing.List`

Comment (pereman2) [reply omegaconf/listconfig.py:73]: In python 3.7 or above it works perfectly fine, in the other hand python 3.6 does not.

Comment (pereman2) [reply omegaconf/listconfig.py:75]: I changed the code so it doesn't allow things like List[None] that I missed.

Comment (omry) [reply omegaconf/listconfig.py:75]: I am still not seeing a test that verifies this.

File: omegaconf/_utils.py (hunk 7 above)
Comment (omry) [inline omegaconf/_utils.py:468]: why is this needed for dict and not for list?

Comment (pereman2) [reply omegaconf/_utils.py:468]: Because is get_dict_key_value_types(), list has its own get_list_element_type()

Comment (omry) [reply omegaconf/_utils.py:468]: I know.  I am not seeing any changes to get_list_element_type(). This problem looks like it effected both List and Dict, so how come there is a fix in get_dict_key_value_types() and not in get_list_element_type()?

File: omegaconf/listconfig.py (hunk 8 above)
Comment (omry) [reply omegaconf/listconfig.py:75]: Let me re-cap one more time:
I asked about a test for pickling Structured Configs with annotations like List[int] and Dict[str, int].
The reason is that I remember some issues.

You said something like:
There was an issue with Python 3.6 but I worked around it.
I asked for the same test again, to be sure that your work around works.

This is not specifically about this PR, but as we are changing the internal representation of DictConfig and ListConfig we need to be sure that we don't introduce regressions.

File: tests/structured_conf/data/dataclasses.py [hunk 9]
```diff
@@ -475,3 +475,15 @@ class MissingStructuredConfigField:
 class ListClass:
     list: List[int] = field(default_factory=lambda: [])
     tuple: Tuple[int, int] = field(default_factory=lambda: (1, 2))
+
+
+@dataclass
+class GenericList:
+    list: List = field(default_factory=lambda: [])  # type: ignore
```
Comment (omry) [inline tests/structured_conf/data/dataclasses.py:482]: put something in the list, like [1,2,3] and verify the content in your test as well.
same for the dict.

File: omegaconf/_utils.py (hunk 7 above)
Comment (pereman2) [reply omegaconf/_utils.py:468]: Because the way get_list_element_type() is implemented returns None(which should).

File: omegaconf/listconfig.py (hunk 8 above)
Comment (pereman2) [reply omegaconf/listconfig.py:75]: I'm going to push the test + the workaround for python3.6 ( I had to change the implementation of __getstate__ on basecontainer. I don't know if that's the best way to do it.

File: omegaconf/basecontainer.py [hunk 10]
```diff
@@ -88,7 +88,19 @@ def __repr__(self) -> str:
...
+        dict_ = copy.copy(self.__dict__)
+        if sys.version_info < (3, 7, 0):  # pragma: no cover
+            ref_type = dict_["_metadata"].ref_type
+            if isinstance(self, DictConfig):
+                if ref_type is not Any:
+                    dict_["_metadata"].ref_type = Dict
+            else:
+                if ref_type is not Any:
+                    dict_["_metadata"].ref_type = List
+
```
Comment (omry) [inline omegaconf/basecontainer.py:102]: Please explain.
looks like you are changing the ref_type on serialization. I don't see why this makes sense.

File: omegaconf/listconfig.py (hunk 8 above)
Comment (omry) [reply omegaconf/listconfig.py:75]: let's chat.

File: tests/structured_conf/test_structured_config.py [hunk 11]
```diff
@@ -750,6 +753,45 @@ def test_assign_wrong_type_to_dict(self, class_type: str, value: Any) -> None:
...
+    def test_pickle_dict(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.GenericDict)
+        with tempfile.TemporaryFile() as fp:
+            pickle.dump(cfg, fp)
+            fp.flush()
+            fp.seek(0)
+            cfg2 = pickle.load(fp)
+            assert cfg == cfg2
+
```
Comment (omry) [inline tests/structured_conf/test_structured_config.py:794]: this belongs with the serialization tests (dedicated test file).

File: omegaconf/basecontainer.py [hunk 12]
```diff
@@ -88,10 +88,39 @@ def __repr__(self) -> str:
 
     # Support pickle
     def __getstate__(self) -> Dict[str, Any]:
-        return self.__dict__
+        from omegaconf import DictConfig
+
+        dict_ = copy.copy(self.__dict__)
+        if sys.version_info < (3, 7, 0):  # pragma: no cover
```
Comment (omry) [inline omegaconf/basecontainer.py:94]: Let's do this unconditionally (for all Python versions).
By conditioning it, you end up with two different file formats, depending on the version that saved.
This makes it tricky if you save in 3.6 and load in 3.7.

File: omegaconf/basecontainer.py [hunk 13]
```diff
@@ -88,10 +88,39 @@ def __repr__(self) -> str:
...
+        return dict_
 
     # Support pickle
     def __setstate__(self, d: Dict[str, Any]) -> None:
+        from omegaconf import DictConfig
+
+        def none_as_any(input_: Any) -> Any:  # pragma: no cover
+            return Any if input_ is None else input_
+
+        if sys.version_info < (3, 7, 0):  # pragma: no cover
```
Comment (omry) [inline omegaconf/basecontainer.py:112]: sane here.

File: tests/test_serialization.py [hunk 14]
```diff
@@ -137,6 +150,44 @@ def test_pickle_list() -> None:
...
+
+@pytest.mark.parametrize(  # type: ignore
+    "input_,node,element_type,key_type,optional,output",
+    [
+        (GenericList, "list", None, None, False, List[Any]),
+        (OptGenericList, "opt_list", None, None, True, List[Any]),
+        (GenericDict, "dict", None, None, False, Dict[Union[str, Enum], Any]),
+        (OptGenericDict, "opt_dict", None, None, True, Dict[Union[str, Enum], Any]),
+        (SubscriptedDict, "dict", int, str, False, Dict[str, int]),
+        (SubscriptedList, "list", int, None, False, List[int]),
```
Comment (omry) [inline tests/test_serialization.py:166]: Add some regular ListConfig and DictConfig with specified key and element types.

File: omegaconf/_utils.py [hunk 15]
```diff
@@ -496,7 +496,12 @@ def valid_value_annotation_type(type_: Any) -> bool:
 
 
 def _valid_dict_key_annotation_type(type_: Any) -> bool:
-    return type_ is None or issubclass(type_, str) or issubclass(type_, Enum)
+    return (
```
Comment (omry) [inline omegaconf/_utils.py:499]: Was this change needed to support the pickling DictConfig and ListConfig?

File: omegaconf/basecontainer.py [hunk 16]
```diff
@@ -88,39 +88,41 @@ def __repr__(self) -> str:
...
-                if ref_type is not Any:
-                    dict_["_metadata"].ref_type = Dict
-            else:
-                if ref_type is not Any:
-                    dict_["_metadata"].ref_type = List
+        ref_type = dict_["_metadata"].ref_type
+        if isinstance(self, DictConfig):
+            if ref_type is not Any:
+                key_type, element_type = _utils.get_dict_key_value_types(ref_type)
+                print(ref_type, key_type, element_type)
```
Comment (omry) [inline omegaconf/basecontainer.py]: You are using print way too much. :)
I do recommend that you install pycharm and learn to use the IDE debugger.

File: omegaconf/basecontainer.py [hunk 17]
```diff
@@ -88,39 +88,41 @@ def __repr__(self) -> str:
...
+                if key_type == Union[str, Enum]:
+                    key_type = None
```
Comment (omry) [inline omegaconf/basecontainer.py:99]: Is this decoded properly on the way out? looks like it will become `Dict[Any, ...]`.
The reason you missed it is because you added logic that is not covered by your tests.

File: omegaconf/basecontainer.py [hunk 18]
```diff
@@ -88,39 +88,41 @@ def __repr__(self) -> str:
...
-            else:
-                if ref_type is not Any:
-                    d["_metadata"].ref_type = List[element_type]  # type: ignore
+        element_type = none_as_any(d["_metadata"].element_type)
+        key_type = none_as_any(d["_metadata"].key_type)
+        if isinstance(self, DictConfig):
+            d["_metadata"].ref_type = Dict[key_type, element_type]  # type: ignore
+        else:
+            d["_metadata"].ref_type = List[element_type]  # type: ignore
+        print(d["_metadata"].ref_type)
```
Comment (omry) [inline omegaconf/basecontainer.py]: remove print.

File: omegaconf/_utils.py (hunk 15 above)
Comment (pereman2) [reply omegaconf/_utils.py:499]: Yes.

File: omegaconf/basecontainer.py (hunk 17 above)
Comment (pereman2) [reply omegaconf/basecontainer.py:99]: Internally is Dict[Any, Any] but it behaves like Dict[Union[str, Enum], Any].

Comment (pereman2) [reply omegaconf/basecontainer.py:99]: If you do get_ref_type it returns the `Dict[Union[str, Enum], Any]`

File: omegaconf/basecontainer.py (hunk 16 above)
Comment (pereman2) [reply omegaconf/basecontainer.py]: When i know where is the problem i tend to use prints because I find it faster.

File: tests/test_serialization.py [hunk 19]
```diff
@@ -136,6 +148,90 @@ def test_pickle_list() -> None:
...
+            ListConfig(content=[1, 2], ref_type=List[int]),
+            None,
+            int,
+            Any,
+            True,
+            Optional[List[int]],
+        ),
+    ],
+)
+def test_pickle_generic(
```
Comment (omry) [inline tests/test_serialization.py:206]: did you move this function?
can you not do it, it makes it harder to review.

Comment (pereman2) [reply tests/test_serialization.py:206]: Sorry, rebasing I missed that test_pickle_generic was duplicated and I had to remove one of them. It looks like I moved it. I'm fixing it.

File: tests/test_serialization.py [hunk 20]
```diff
@@ -136,6 +148,90 @@ def test_pickle_list() -> None:
...
+            Optional[List[int]],
+        ),
+        (
+            ListConfig(content=[1, 2], ref_type=List[int]),
+            None,
+            int,
+            Any,
+            True,
+            Optional[List[int]],
+        ),
```
Comment (pereman2) [inline tests/test_serialization.py:203]: Im thinking of removing this test cases since ListConfigs and DictConfigs are never initialised like this.

Comment (omry) [reply tests/test_serialization.py:203]: This is emulating how they are initialized with Structured Configs. I think it's fine.

Comment (pereman2) [reply tests/test_serialization.py]: Okay, I've modified this tests since ref_type is never passed alone without extracting the key_type and element_type(we get key_type and element_type and then initialise with it). 
I've removed resolve ref_type since it doesn't make sense after all previous PR and can be solved without it and, because of this, I modified tests to get ref_type with `_utils.get_ref_type`.

Comment (omry) [reply tests/test_serialization.py]: Instead of removing the tests, just pass the additional required information in (key_type and element_type).

Comment (pereman2) [reply tests/test_serialization.py]: Yes, I did that but that made duplicated test cases which I removed.

File: tests/test_serialization.py (hunk 19 above)
Comment (omry) [reply tests/test_serialization.py:195]: Still not moved back.

**Review by omry — CHANGES_REQUESTED** [review]
Please move test_pickle_generic to it's original place so it will be possible to review what changed there.

**Comment by pereman2** [comment]
See ae5aca9

File: tests/test_utils.py (hunk 2 above)
Comment (omry) [reply tests/test_utils.py:214]: We said we will remove this fixture.

Can you remove?
also, can you go over the whole diff to be sure it all makes sense? I feel like I keep finding resurrected things.

**Comment by pereman2** [comment]
Something missing?

**Comment by omry** [comment]
Time. There is never enough time.
My weekend also started. and I am also taking Monday off.
:)

**Comment by pereman2** [comment]
Okay have fun!

File: tests/test_serialization.py [hunk 21]
```diff
@@ -136,6 +148,9 @@ def test_pickle_list() -> None:
         fp.seek(0)
         c1 = pickle.load(fp)
         assert c == c1
+        assert get_ref_type(c1) == Optional[List[Any]]
+        assert c1._metadata.element_type is Any
+        assert c1._metadata.optional is True
```
Comment (omry) [inline tests/test_serialization.py:153]: You are testing this in the following test, right?
should those two tests be merged?

File: tests/test_utils.py (hunk 2 above)
Comment (omry) [reply tests/test_utils.py:214]: You missed my request again.

**Review by omry — CHANGES_REQUESTED** [review]
remove the fixture you added.

File: tests/__init__.py [hunk 22]
```diff
@@ -146,11 +146,32 @@ class Package:
...
+@dataclass
+class OptGenericList:
+    opt_list: Optional[List] = None  # type: ignore
+
+
+@dataclass
+class GenericList:
+    list: List = field(default_factory=lambda: [1, 2])  # type: ignore
+
+
```
[the remark this answers is not in the document]
Comment (pereman2) [reply tests/__init__.py:158]: Hmmm Dict[kt, vt] is a generic type therefore I should't have used Generic for Dict alone. 
Maybe we should call it TypelessDict? I haven't seen a way to name this in the docs.

[the remark this answers is not in the document]
Comment (omry) [reply tests/__init__.py:158]: UntypedDict, UntypedList ?

File: NEWS.md [hunk 23]
```diff
@@ -1,3 +1,10 @@
+## 2.0.5 (2020-11-11)
+
+
+### Bug Fixes
+
+- Fix bug where interpolations were unnecessarily resolved during merge ([#431](https://github.com/omry/omegaconf/issues/431))
+
```
Comment (omry) [inline NEWS.md:7]: Please clean you diff not include changes from other diffs. (rebase, don't merge).

**Review by omry — CHANGES_REQUESTED** [review]

File: tests/structured_conf/test_structured_config.py [hunk 24]
```diff
@@ -811,6 +812,23 @@ def test_recursive_list(self, class_type: str) -> None:
         cfg = OmegaConf.structured(o)
         assert cfg == {"d": [{"d": "???"}, {"d": "???"}]}
 
+    def test_create_generic_dict(self, class_type: str) -> None:
```
Comment (omry) [inline tests/structured_conf/test_structured_config.py:815]: create_untyped_dict ?

File: tests/structured_conf/test_structured_config.py [hunk 25]
```diff
@@ -811,6 +812,23 @@ def test_recursive_list(self, class_type: str) -> None:
...
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.UntypedDict)
+        dt = Dict[Union[str, Enum], Any]
+        assert _utils.get_ref_type(cfg, "dict") == dt
+        assert _utils.get_ref_type(cfg, "opt_dict") == Optional[dt]
+        assert cfg.dict == {"foo": "var"}
+        assert cfg.opt_dict is None
+
+    def test_create_generic_list(self, class_type: str) -> None:
```
Comment (omry) [inline tests/structured_conf/test_structured_config.py:824]: create_untyped_list?

File: tests/test_create.py [hunk 26]
```diff
@@ -198,3 +199,17 @@ def test_create_unmodified_loader() -> None:
...
+
+    cfg = ListConfig(ref_type=List, content=[])
+    assert get_ref_type(cfg) == Optional[List[Any]]
+
+
+def test_create_generic_dict() -> None:
+    from omegaconf._utils import get_ref_type
+
+    cfg = DictConfig(ref_type=Dict, content={})
+    assert get_ref_type(cfg) == Optional[Dict[Union[str, Enum], Any]]
```
Comment (omry) [inline tests/test_create.py:215]: ditto.

File: tests/test_utils.py [hunk 27]
```diff
@@ -211,6 +211,7 @@ def test_get_structured_config_data_illegal_value(test_cls: Any) -> None:
     assert d["x"] == IllegalType()
 
 
+@pytest.fixture  # type: ignore
```
Comment (omry) [inline tests/test_utils.py:214]: So, I don't understand why this is here.
I asked for it to be removed 4 times or so.
am I missing anything?

**Review by omry — CHANGES_REQUESTED** [review]

File: tests/test_utils.py (hunk 27 above)
Comment (pereman2) [reply tests/test_utils.py:214]: Oh, I removed it before messing the rebase sorry.

**Review by omry — APPROVED** [review]
Looking good, thanks!
